proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,250 @@
+| `struct Foo { RWBuffer<float> f; }; Foo foo : register(c2);`| "warning: 'c' register type should only be used for basic types. Disable with --Wno-disallow-legacy-binding-rules" |
+| `RWBuffer<float> f : register(c3);`| "error: invalid register specification, expected 'u' binding" |
+
+
+## Detailed design
+
+All the compiler has to work with is a Decl object and the context the decl appears in.
+From this information, the first goal is to set a specific set of flags that can fully inform
+the right diagnostic to emit. The first group of flags are the decl type flags, which
+are either `basic`, `resource`, `udt`, or `other`. `basic` refers to a numeric variable
+type. `other` refers to a type that cannot possibly be resource, because it lacks the 
+resource attribute.<br>
+The next group of flags are the resource class flags,
+which are `srv`, `uav`, `cbv`, or `sampler`.<br>
+The final flag is `default_globals`, which indicates whether or not the decl appears
+inside the $Globals scope.<br><br>
+
+From the Decl object, first determine if the decl is a cbuffer or tbuffer decl. If so,
+we know the decl is a `resource` and can infer the resource class (`cbv` or `srv`
+respectively). Otherwise, the decl is classified according to the underlying canonical 
+type.<br>
+If the decl is implicit, then check if the decl is a vector, matrix, or otherwise numeric type,
+and if so, we know that the decl is a `basic` type. If the decl has a resource attribute, 
+we can set the `resource` flag, and can also obtain the resource class and set the resource 
+class flag accordingly. Otherwise, we know the decl is not a resource and not basic, and so
+the `other` flag is set.<br>
+If the decl is not implicit, then we check if it is a struct or class type, and if so, the
+decl is a UDT, so the `udt` flag is set. We then recurse through the members of the UDT and
+collect information on what types of resources are contained by the UDT (uav, cbv, srv, or
+sampler). The corresponding resouce class flags are set. If the decl is not a struct or class
+type, then it is a basic type, and so the `basic` flag is set.<br>
+The last step is to simply check if the decl is in the `$Globals` scope. If so, then the
+`default_globals` flag will be set.<br><br>
+
+Now enough information has been gathered to determine the right diagnostics to emit, if any.
+Firstly, if `other` is set, then the variable type is not valid for the given register type.
+So, we can emit this diagnostic: `error: '%0' is an invalid resource type for register type '%1'`
+If the `resource` flag is set and the `udt` flag is not set, then we have to check the 
+register statement, and the resource class flag, and verify that the register type matches the 
+resource class flag. If not, we emit this diagnostic: 
+`error: %select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used.`

```

Comment:
`resource` and `udt` should be mutually exclusive.  Either the type is a single resource type and `resource` is set, or it's a user defined type (that could potentially contain resources) and `udt` is set.

```suggestion
If the `resource` flag is set, check the register type against the resource class flag.
If they do not match, emit this diagnostic:
`error: %select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used.`

```